src/main/java/de/zalando/aruha/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -0,0 +1,366 @@
+package de.zalando.aruha.nakadi.service.subscription.state;
+
+import de.zalando.aruha.nakadi.service.EventStream;
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZKSubscription;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class StreamingState extends State {
+    private ZKSubscription topologyChangeSubscription;
+    private Consumer<String, String> kafkaConsumer;
+    private final Map<Partition.PartitionKey, PartitionData> offsets = new HashMap<>();
+    // Maps partition barrier when releasing must be completed or stream will be closed.
+    // The reasons for that if there are two partitions (p0, p1) and p0 is reassigned, if p1 is working
+    // correctly, and p0 is not receiving any updates - reassignment won't complete.
+    private final Map<Partition.PartitionKey, Long> releasingPartitions = new HashMap<>();
+    private boolean pollPaused = false;
+    private long lastCommitMillis;
+    private long commitedEvents = 0;
+    private long sentEvents = 0;
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamingState.class);
+
+    @Override
+    public void onEnter() {
+        // Create kafka consumer
+        this.kafkaConsumer = getKafka().createKafkaConsumer();
+        // Subscribe for topology changes.
+        this.topologyChangeSubscription = getZk().subscribeForTopologyChanges(() -> addTask(this::topologyChanged));
+        // and call directly
+        reactNoTopologyChange();
+        addTask(this::pollDataFromKafka);
+        scheduleTask(this::checkBatchTimeouts, getParameters().batchTimeoutMillis, TimeUnit.MILLISECONDS);
+
+        if (null != getParameters().streamTimeoutMillis) {
+            scheduleTask(this::shutdownGracefully, getParameters().streamTimeoutMillis, TimeUnit.MILLISECONDS);
+        }
+
+        this.lastCommitMillis = System.currentTimeMillis();
+        scheduleTask(this::checkCommitTimeout, getParameters().commitTimeoutMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private void checkCommitTimeout() {
+        if (!isCurrent()) {
```

Comment:
Remember current state while scheduling task (and this can be easily done) - decision.